### PR TITLE
API: Enable `pydata/sparse` input for csgraph module

### PR DESCRIPTION
**Pull request by mtsokol**
Hi!
This PR adds simple support for `pydata/sparse` inputs to `scipy.sparse.csgraph`, similarly to `scipy.sparse.linalg`, where `pydata/sparse` is supported. 

Here, for now, by "support" I mean converting to `scipy` sparse format.

### CODE REVIEW

**Comment by lucascolley** [comment]
I assume we'll want to test this like https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/tests/test_pydata_sparse.py.

**Comment by rgommers** [comment]
> (do we have coverage of that in CI for `sparse.linalg` currently?).

Great question - I think we had it at some point, and silently lost it somewhere in a CI reshuffle.

We probably shouldn't add it to the test dependencies in `requirements/test.txt` because the dependency on Numba is a bit fragile at times (Python 3.12 support requires `--pre` right now). but we can add a simply `pip install sparse` to one Linux GHA CI job. I suggest the `prerelease_deps_...` one in `.github/workflows/linux_meson.yml`.

**Comment by mtsokol** [comment]
@rgommers @lucascolley Thanks for a feedback!

I re-enabled `test_pydata_sparse.py` for `scipy.sparse.linalg` (used `pip install sparse` to get the dependency) and now it runs in CI (one test is failing with atol overrun).

I also added `test_pydata_sparse.py` for `scipy.sparse.csgraph` submodule to verify that each public function accepts pydata/sparse input.

By "accepts" it means converting pydata/sparse object to a `scipy.sparse` matrix with a desired format (some functions require `csr` and some don't care). If the result type is sparse, it's always a `scipy.sparse` matrix instance, no wrapping back to pydata/sparse.

**Comment by lucascolley** [comment]
Sounds good!

> If the result type is sparse, it's always a scipy.sparse matrix instance, no wrapping back to pydata/sparse.

How is this going to work long-term? I suppose we are still waiting for `scipy.sparse` arrays and `pydata/sparse` arrays to develop as we move away from matrices, but is the plan also to have 'input type in == input type out' eventually?

File: scipy/sparse/_sputils.py [hunk 1]
```diff
@@ -388,6 +390,27 @@ def is_pydata_spmatrix(m) -> bool:
     return base_cls is not None and isinstance(m, base_cls)
 
 
+def convert_pydata_sparse_to_scipy(
```
Comment (perimosocordiae) [inline scipy/sparse/_sputils.py:393]: I'd prefer to have a function that takes a single input and returns a single output. In cases where we need to convert multiple objects, we can call the function multiple times.

File: scipy/sparse/_sputils.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 """ Utility functions for sparse matrix module
 """
 
+from collections.abc import Iterable
 import sys
+from typing import Any, Literal as L, Optional, Union
```
Comment (perimosocordiae) [inline scipy/sparse/_sputils.py:6]: Let's avoid aliasing `Literal` for now, unless this is common throughout the rest of the codebase.

File: scipy/sparse/linalg/tests/test_pydata_sparse.py [hunk 3]
```diff
@@ -176,9 +176,9 @@ def test_splu(matrices):
     Pc = sparse_cls(sp.csc_matrix((np.ones(n), (np.arange(n), lu.perm_c))))
     A2 = Pr.T @ lu.L @ lu.U @ Pc.T
 
-    assert_allclose(A2.toarray(), A_sparse.toarray())
+    assert_allclose(A2.todense(), A_sparse.todense())
```
Comment (perimosocordiae) [inline scipy/sparse/linalg/tests/test_pydata_sparse.py:179]: Why change from `.toarray()` to `.todense()` here?

File: scipy/sparse/_sputils.py [hunk 4]
```diff
@@ -388,6 +390,27 @@ def is_pydata_spmatrix(m) -> bool:
...
+    """
+    Convert any pydata/sparse arrays to scipy sparse matrices,
+    pass through anything else.
+    """
+    result = []
+    for arg in args:
+        if is_pydata_spmatrix(arg):
+            arg = arg.to_scipy_sparse()
+            if target_format == "csc":
+                arg = arg.tocsc()
```
Comment (perimosocordiae) [inline scipy/sparse/_sputils.py:405]: You can use `arg.asformat(target_format)` to simplify this.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

**Comment by lucascolley** [comment]
I suppose this should be documented somewhere. Are you planning follow-ups? If not, then something should be added here, although I can't find anything for the existing support in `sparse.linalg` either.

**Comment by mtsokol** [comment]
> I suppose this should be documented somewhere. Are you planning follow-ups?

For sure, I planned a few tweaks for later.

File: scipy/sparse/_sputils.py (hunk 1 above)
Comment (mtsokol) [reply scipy/sparse/_sputils.py:393]: Sure! Now the function accepts single input only.

File: scipy/sparse/_sputils.py (hunk 2 above)
Comment (mtsokol) [reply scipy/sparse/_sputils.py]: Done!

File: scipy/sparse/linalg/tests/test_pydata_sparse.py (hunk 3 above)
Comment (mtsokol) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py:179]: I think `.toarray()` is incorrect. `pydata/sparse` object don't have this function, they provide `.todense()` instead.
`.toarray()` is `scipy.sparse`'s method, so I think it's incorrect here, considering that this test suite isn't even run in the CI in the main branch.

File: scipy/sparse/_sputils.py (hunk 4 above)
Comment (mtsokol) [reply scipy/sparse/_sputils.py]: Done!

**Comment by mtsokol** [comment]
@perimosocordiae I think it's ready for another review.

**Review by perimosocordiae — APPROVED** [review]
Looks good, thanks!

**Comment by rgommers** [comment]
I'm afraid my suggestion to use the `prerelease` job didn't work:
```
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 1.26.3
    Uninstalling numpy-1.26.3:
      Successfully uninstalled numpy-1.26.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
numba 0.58.1 requires numpy<1.27,>=1.22, but you have numpy 2.0.0.dev0 which is incompatible.
```

That will continue to be an issue; the released `numba` version is unlikely to work with a pre-release version of `numpy`. Can you move the install of `sparse` to the `gcc8` job @mtsokol?

File: scipy/sparse/csgraph/tests/test_pydata_sparse.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+try:
+    import sparse
+except Exception:
+    sparse = None
+
+pytestmark = pytest.mark.skipif(sparse is None,
+                                reason="pydata/sparse not installed")
+
+
+msg = "pydata/sparse (0.8) does not implement necessary operations"
```
Comment (rgommers) [inline scipy/sparse/csgraph/tests/test_pydata_sparse.py:18]: If this is still true, can you update the version number to `0.15.1`?

Comment (rgommers) [reply scipy/sparse/csgraph/tests/test_pydata_sparse.py:18]: Also in the linalg test file

Comment (mtsokol) [reply scipy/sparse/csgraph/tests/test_pydata_sparse.py]: Yes, the message refers to `DOK` format only. I updated it with `0.15.1`.

**Comment by mtsokol** [comment]
> Can you move the install of `sparse` to the `gcc8` job @mtsokol?

Sure! Now `sparse` is installed in `gcc-8` job.

**Comment by rgommers** [comment]
Now the tests are running as they should. There's one deprecation warning visible:
```
sparse/linalg/tests/test_pydata_sparse.py::test_spilu[COO]
  /opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/sparse/_coo/core.py:215: DeprecationWarning: coords should be an ndarray. This will raise a ValueError in the future.
    warnings.warn(
```
Could you fix that one?

**Comment by mtsokol** [comment]
> Could you fix that one?

@rgommers I started looking at it and it seems that for some functions in `scipy.sparse.linalg` (for this module the support for `sparse` was added ~4 years ago) the result is wrapped in `sparse` class before returning, instead of just returning `scipy`'s sparse matrix class.

Here my understanding is that `sparse` objects get transformed to `scipy`'s sparse matrices when passing to a function but they aren't transformed back to `sparse`. Is it Ok? (Then I could remove this transform on return)

**Comment by rgommers** [comment]
I think the last commit on https://github.com/rgommers/scipy/tree/pydata-sparse-scgraph is the right fix. The code is a little hard to follow, because reconstructing the pydata sparse arrays is done from within the C code in the SuperLU wrapper. But the problem was this construct:
```python
    if is_pydata_spmatrix(A):
        def csc_construct_func(*a, cls=type(A)):
            return cls(csc_matrix(*a))
```
It's trying to reconstruct pydata sparse arrays directly by calling `sparse.COO.__init__`, but it should be calling the SciPy-specific constructor with:
```python
    return cls.from_scipy_sparse(csc_matrix(*a))
```



> the result is wrapped in `sparse` class before returning, instead of just returning `scipy`'s sparse matrix class.

Yes, this is the expected behavior I'd say. The code in `sparse.linalg` is already doing that, and we want that "array type out == array type in" to be the general design rule - see gh-18286.

Which means that we should also do that in this PR for `sparse.csgraph` functions.

**Comment by mtsokol** [comment]
@rgommers Thank you! I added changes from your commit and it seems that the warning is gone in `gcc-8`.

I also updated un/wrapping code so that "array type out == array type in" for `pydata/sparse` for `csgraph` submodule (and added `isinstance` checks in test).

It's ready from my side.

File: scipy/sparse/csgraph/tests/test_pydata_sparse.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+
+def test_dijkstra(graphs):
+    A_dense, A_sparse = graphs
+    func = spgraph.dijkstra
+
+    actual = func(A_sparse)
+    desired = func(sp.csc_matrix(A_dense))
+
+    assert_equal(actual, desired)
```
Comment (ivirshup) [inline scipy/sparse/csgraph/tests/test_pydata_sparse.py:85]: This test shares the exact same structure as a number of other cases. I would suggest that `func` is used a parameter and test code is deduplicated, like:

```suggestion
@pytest.mark.parameterize(
    "func",
    [
        spgraph.shortest_path,
        spgraph.dijkstra,
        spgraph.floyd_warshall,
        spgraph.bellman_ford,
        spgraph.johnson,
        spgraph.reverse_cuthill_mckee,
        spgraph.maximum_bipartite_matching,
        spgraph.structural_rank,
    ]
)
def test_csgraph_equiv(func, graphs):
    A_dense, A_sparse = graphs

    actual = func(A_sparse)
    desired = func(sp.csc_matrix(A_dense))

    assert_equal(actual, desired)
```

Comment (rgommers) [reply scipy/sparse/csgraph/tests/test_pydata_sparse.py:85]: This looks like a good suggestion to me. 

Other than that, this PR looks ready to merge.

Comment (mtsokol) [reply scipy/sparse/csgraph/tests/test_pydata_sparse.py]: Right, they're quite similar - I a parametrized `test_csgraph_equiv`.

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Nice work @mtsokol, and thanks for the reviews @perimosocordiae and @ivirshup!
